manpage/jetbrains-macos-cli.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "SCRIPT_NAME" "1" "December 2023" "JetBrains" ""
.
.SH "NAME"
\fBscript_name\fR \- Command\-line interface for ide_name
.
.SH "SYNOPSIS"
\fBscript_name\fR [\fIfile\-or\-dir\fR \fI\.\.\.\fR] [\fBnosplash\fR] [\fBdontReopenProjects\fR] [\fBdisableNonBundledPlugins\fR]
.
.P
\fBscript_name \-\-wait\fR \fIfile\-or\-dir\fR [\fBnosplash\fR] [\fBdisableNonBundledPlugins\fR]
.
.P
\fBscript_name diff\fR \fIfile1\fR \fIfile2\fR [\fIfile3\fR]
.
.P
\fBscript_name merge\fR \fIfile1\fR \fIfile2\fR [\fIcommon\-base\-file\fR] \fIoutput\-file\fR
.
.P
\fBscript_name format\fR \fIfile\-or\-dir\fR \fI\.\.\.\fR [\fB\-m\fR|\fB\-mask\fR \fImasks\fR] [\fB\-r\fR|\fB\-R\fR] [\fB\-s\fR|\fB\-settings\fR \fIsettings\-file\fR] [\fB\-allowDefaults\fR] [\fB\-charset\fR \fIcharset\fR] [\fB\-d\fR|\fB\-dry\fR]
.
.P
\fBscript_name inspect\fR \fIproject\-dir\fR \fIinspection\-profile\-xml\-file\fR \fIoutput\-dir\fR [\fB\-changes\fR] [\fB\-d\fR \fIsubdirectory\fR] [\fB\-format\fR \fBxml\fR|\fBjson\fR|\fBplain\fR] [\fB\-v0\fR|\fB\-v1\fR|\fB\-v2\fR]
.
.P
\fBscript_name installPlugins\fR \fIplugin\-id\fR \fI\.\.\.\fR [\fIrepository\-url\fR \fI\.\.\.\fR]
.
.SH "DESCRIPTION"
Use ide_name features from the command line: open files and projects, view diffs, merge files, format code, inspect code, and install plugins\.
.
.SH "LAUNCH THE IDE, OPEN FILES/DIRECTORIES"
\fBscript_name\fR [\fIfile\-or\-dir\fR \fI\.\.\.\fR] [\fIoption\fR \fI\.\.\.\fR]
.
.P
Without any arguments, ide_name launches, as though its icon was clicked\. For instance, if a project was open when the IDE was last closed, it is reopened\.
.
.P
When files or directories are passed, ide_name opens them instead of restoring its previous state\. When you specify the path to a file, ide_name opens it in the LightEdit mode, unless it belongs to a project that is already open or there is special logic to automatically open or create a project (for example, in case of Maven or Gradle files)\. If you specify a directory with an existing project, ide_name opens this project\. If you open a directory that is not part of a project, ide_name adds the \fB\.idea\fR directory to it, making it a project\.
.
.P
\fBOptions\fR
.
.P
Options may be given anywhere on the command line, before, after or in between other arguments\.
.
.TP
\fB\-\-wait\fR | \fB\-wait\fR | \fB\-w\fR
Waits for the file or directory to be closed before returning to the command prompt\. For example, if you open file\.txt with \fBscript_name \-\-wait file\.txt\fR, the shell will be waiting until file\.txt is closed in ide_name\. You may pass only a single file or directory when using this option\.
.
.TP
\fB\-\-nosplash\fR | \fB\-nosplash\fR | \fBnosplash\fR
Do not show the splash screen when loading ide_name\.
.
.TP
\fB\-\-dontReopenProjects\fR | \fB\-dontReopenProjects\fR | \fBdontReopenProjects\fR
Do not reopen projects, and instead show the welcome screen\. This can help if a project that was open crashes ide_name\.
.
.TP
\fB\-\-disableNonBundledPlugins\fR | \fB\-disableNonBundledPlugins\fR | \fBdisableNonBundledPlugins\fR
Do not load manually installed plugins\. This can help if a plugin that you installed crashes ide_name\. You will be able to start the IDE and disable or uninstall the problematic plugin\.
.
.P
Note that some of these options are accepted without leading dashes\. To open a directory whose name matches one of these options in ide_name, pass the directory name with a path component, like \fB~/nosplash\fR or \fB\./nosplash\fR\.
.
.SH "COMPARE FILES"
\fBscript_name diff\fR \fIfile1\fR \fIfile2\fR [\fIfile3\fR]
.
.P
Open the Differences Viewer to compare two or three files from the command line\. For example, you can compare the current version of a file with its backup, or your local copy of a file with its copy from the remote repository or its copy from another branch\.
.
.P
You can compare files of any types, including binaries and \.jar files\. The Differences Viewer provides a powerful editor that enables code completion, live templates, and other features\.
.
.SH "MERGE FILES"
\fBscript_name merge\fR \fIfile1\fR \fIfile2\fR [\fIcommon\-base\-file\fR] \fIoutput\-file\fR
.
.P
Open the Merge dialog to perform a two\-way or three\-way merge\.
.
.P
To perform a three\-way merge, you need to specify paths for two modified versions of a file, the base revision (a common origin of both modified versions), and the output file to save merge results\.
.
.P
Don\'t specify the optional base revision if you want to treat the current contents of the output file as the common origin\. In this case, if the output is an empty file, this essentially becomes a two\-way merge\.
.
.SH "FORMAT SOURCE CODE"
\fBscript_name format\fR \fIfile\-or\-dir\fR \fI\.\.\.\fR [\fIoption\fR \fI\.\.\.\fR]
.
.P
ide_name lets you reformat your code according to the requirements you\'ve specified in your current code style scheme or the \fB\.editorconfig\fR file\. If anything is not defined in \fB\.editorconfig\fR, it is taken from the project settings\.
.
.P
The command\-line formatter launches an instance of ide_name in the background and applies the formatting\. It will not work if another instance of ide_name is already running\. In this case, you can perform code style formatting from the running instance\. Use the command\-line formatter for automated regular maintenance of a large codebase with many contributors to ensure a consistent coding style\.
.
.P
To be able to format files, install and enable plugins with support for the corresponding file types in ide_name (for example, the Shell Script plugin to format shell script files)\.
.
.P
\fBOptions\fR
.
.P
Options may be given anywhere on the command line, before, after or in between other arguments\.
.
.TP
\fB\-m\fR \fImasks\fR | \fB\-mask\fR \fImasks\fR
Only process files that match one of the specified masks\. Pass a comma\-separated list of masks, using * and ? wildcards, matching any string and any single character, respectively\.
.
.TP
\fB\-r\fR | \fB\-R\fR
Process specified directories recursively\.
.
.TP
\fB\-s\fR \fIsettings\-file\fR | \fB\-settings\fR \fIsettings\-file\fR
Specify the code style settings file to use for formatting\. This can be either a file containing exported code style settings (open the \fBEditor\fR > \fBCode Style\fR page of the IDE settings, click the gear icon, and select Export), or the \fB\.idea/codeStyles/Project\.xml\fR file stored in your project directory\.
.
.IP
The formatter also looks for \fB\.editorconfig\fR files in the parent directories and applies them for formatting on top of the ide_name code style settings\. In this case, if formatting settings from EditorConfig overlap with the settings from your code style scheme, ide_name will use the settings from EditorConfig\. The remaining settings will be taken from your code style scheme\.
.
.IP
If there is a project in one of the parent folders, its settings will be used implicitly as well as EditorConfig\.
.
.IP
Files outside a project will be skipped unless you pass either this option or the \fB\-allowDefaults\fR option\.
.
.TP
\fB\-allowDefaults\fR
Use the default code style settings when the style is not defined for a file: when \fB\-s\fR/\fB\-settings\fR was not passed and the file does not belong to any project\. Otherwise, the file will be skipped and not formatted\.
.
.TP
\fB\-charset\fR \fIcharset\fR
Preserve encoding and enforce the character set for reading and writing source files\. For example: \fB\-charset ISO\-8859\-15\fR
.
.IP
This option is useful if the command\-line formatter cannot correctly process special letters in a source file\.
.
.TP
\fB\-d\fR | \fB\-dry\fR
Dry\-run/validation mode\. The formatter will exit with a non\-zero status if it would modify any files, or zero if all files are already formatted\.
.
.P
\fBExamples\fR
.
.TP
Format two specific files from the ~/MyProject/src directory using the project\'s code style and EditorConfig settings:
.
.IP
\fBscript_name format ~/MyProject/src/foo\.html ~/MyProject/src/goo\.html\fR
.
.TP
Format all files in the ~/Data/src directory, recursively including files in all subdirectories, using the default code style settings:
.
.IP
\fBscript_name format ~/Data/src \-r\fR
.
.TP
Format all the \.xml and \.html files in the current directory (not subdirectories), using code style settings from ~/Data/settings\.xml:
.
.IP
\fBscript_name format \-settings ~/Data/settings\.xml \-mask "*\.xml,*\.html" \.\fR
.
.SH "INSPECT SOURCE CODE"
\fBscript_name inspect\fR \fIproject\-dir\fR \fIinspection\-profile\-xml\-file\fR \fIoutput\-dir\fR [\fIoption\fR \fI\.\.\.\fR]
.
.P
Run all the configured inspections for a project, and store the results as an XML, JSON, or plain text file with a report\.
.
.P
ide_name runs code inspections to find and highlight syntax errors, dead code, possible bugs, bad coding style, and other problems\. You can run all the configured inspections for a specific project and store the results as an XML, JSON, or plain text file with a report\.
.
.P
An inspection profile is an XML file that specifies enabled inspections and their options\. By default, project inspection profiles are stored in the project directory: \fB\.idea/inspectionProfiles\fR\. Global IDE inspection profiles are stored in the IDE\'s configuration directory\'s \fBinspection\fR subdirectory: \fB~/Library/Application Support/JetBrains/<product><version>/inspection\fR\.
.
.P
You can open the generated file with inspection results in ide_name or any text editor as a regular file\. However, it is more convenient to analyze it using the Offline View of the Problems tool window: Open the project for which you ran the inspection, go to Code | Analyze Code | View Offline Inspection Results, then select the directory with the inspection results and click OK\.
.
.P
The command\-line inspector launches an instance of ide_name in the background where it runs the inspections\. It will not work if another instance of ide_name is already running\. In this case, you can use inspections in the running instance\. Use the command\-line inspector for automated regular code analysis of large projects with many contributors\.
.
.P
Note that code inspections rely on a properly defined project SDK\.
.
.P
\fBOptions\fR
.
.P
Options may be given anywhere on the command line, before, after or in between other arguments\.
.
.TP
\fB\-changes\fR
Run inspections only on local uncommitted changes\.
.
.TP
\fB\-d\fR \fIsubdirectory\fR
The full path of a subdirectory to inspect, rather than inspecting the whole project\.
.
.TP
\fB\-format\fR \fBxml\fR|\fBjson\fR|\fBplain\fR
The format of the output file containing inspection results: XML (default), JSON, or plain text\.
.
.TP
\fB\-v0\fR | \fB\-v1\fR | \fB\-v2\fR
Set the verbosity level of the output\. \fB\-v0\fR is low verbosity (default), \fB\-v1\fR is medium, and \fB\-v2\fR is maximum verbosity\.
.
.P
\fBExample\fR
.
.TP
Inspect the \fBsrc\fR directory of the \fB~/MyProject\fR project using the project\'s \fBMyProfile\.xml\fR inspection profile, with maximum verbosity, and save results to an XML file in the \fB~/MyProject/InspectionResults\fR directory:
.
.IP
\fBscript_name inspect \-d ~/MyProject/src ~/MyProject ~/MyProject/\.idea/inspectionProfiles/MyProfile\.xml \-v2 ~/MyProject/InspectionResults\fR
.
.SH "INSTALL PLUGINS"
\fBscript_name installPlugins\fR \fIplugin\-id\fR \fI\.\.\.\fR [\fIrepository\-url\fR \fI\.\.\.\fR]
.
.P
Install plugins from the JetBrains Marketplace (\fIhttps://plugins\.jetbrains\.com\fR), or from a custom plugin repository\.
.
.P
To install a plugin, you must know its Plugin ID\. The developer of a plugin specifies this unique identifier for the plugin in its plugin configuration file\. If it is a public plugin, you can find its ID on the plugin\'s page in JetBrains Marketplace: on the Versions tab, click the relevant version to view the Plugin ID and other details\.
.
.P
Quit IntelliJ IDEA before running this command!
.
.P
Installing plugins means putting them in the plugins directory\. If you run this command on a fresh IntelliJ IDEA installation, the IDE will not suggest importing the settings from the previous installation when you launch it for the first time, because the settings will no longer be empty\.
.
.P
\fBExamples\fR
.
.TP
Install the Grazie Lite plugin:
.
.IP
\fBscript_name installPlugins tanvd\.grazi\fR
.
.TP
Install a plugin with the ID \fBcom\.example\.myplugin\fR from a custom repository \fBhttp://plugins\.example\.com:8080/updatePlugins\.xml\fR:
.
.IP
\fBscript_name installPlugins com\.example\.myplugin http://plugins\.example\.com:8080/updatePlugins\.xml\fR
.
.SH "SEE ALSO"
cli_help_url
.\" jetbrains-macos-cli



jetbrains-macos-cli.sh
#!/bin/bash
# Command-line interface for IntelliJ IDEA and various other JetBrains IDEs on macOS.
# jetbrains-macos-cli

ide_app="/Applications/IntelliJ IDEA.app"
ide_binary="idea"
ide_name="IntelliJ IDEA"

if [[ -z $NO_COLOR ]]; then
	bright='\033[0;1m'
	plain='\033[0m'
	light_red='\033[1;31m'
	dark_red='\033[0;31m'
	dim='\033[38;5;240m'
else
	bright=''
	plain=''
	light_red=''
	dark_red=''
	dim=''
fi

# This script's name (or symlink's name, if the script was launched via symlink)
script_name="$(basename "$0")"

# Shows usage info
# See https://www.jetbrains.com/help/idea/working-with-the-ide-features-from-command-line.html
function usage() {
	use "$script_name" \
		"Launch the IDE as though via its icon (restoring previously open projects, etc.)
		Options: nosplash, dontReopenProjects, disableNonBundledPlugins"

	use "$script_name <file-or-dir ...>" \
		"Open arbitrary files and/or directories in the IDE. If you open a directory that's
		not in a project, the IDE will add a .idea directory to it, making it a project.
		Options: nosplash, disableNonBundledPlugins"

	use "$script_name --wait <file-or-dir>" \
		"Open a file or directory, and wait for it to be closed.
		Options: nosplash, disableNonBundledPlugins"

	use "$script_name diff <file1> <file2> [file3]" \
		"Open the diff viewer to compare two or three files."

	use "$script_name merge <file1> <file2> [common-base-file] <output-file>" \
		"Open the Merge dialog to perform a three-way or two-way merge."

	# If you pass a mask, only files that match it will be formatted,
	# even if you explicitly pass a non-matching file explicitly;
	# we don't check for that condition in this script, or consider it a problem
	use "$script_name format <file-or-dir ...>" \
		"Format the given files and/or the files in the given directories, optionally recursively.
		By default, files are formatted with the project's settings; override with -s/-settings.
		Files outside a project are skipped unless -s/-settings or -allowDefaults is passed.
		Options: -m|-mask masks, -r|-R, -s|-settings file.xml, -allowDefaults, -charset set, -d|-dry"
	
	use "$script_name inspect <project-dir> <inspection-profile-xml-file> <output-dir>" \
		"Run all the configured inspections for a project, and store the results in a file.
		Options: -changes, -d subdir, -format xml|json|plain, -v0|-v1|-v2 (verbosity)"

	use "$script_name installPlugins <plugin-id ...> [repository-url ...]" \
		"Install plugins, by default from the JetBrains Marketplace (https://plugins.jetbrains.com)."

	if man -S1 -w "$script_name" &> /dev/null; then
		echo -e $'\n'"${bright}Run 'man ${script_name}' for full usage details${plain}"
	fi
}

# Helper function for usage()
function use() {
	local form=$1; shift
	local description=$1; shift

	echo -e "\n${bright}${form}${plain}"

	local lines line indent="    "
	if [[ -n $description ]]; then
		# shellcheck disable=SC2206
		IFS=$'\n' lines=($description)

		for line in "${lines[@]}"; do
			line="${line#"${line%%[![:space:]]*}"}"  # Remove leading whitespace
			echo "${indent}${line}"
		done
	fi

	# Show any options
	if (( ${#@} > 0 )); then
		echo

		local opt i=0
		for opt; do
			(( i++ ))
			if (( i % 2 )); then
				echo -e "${indent}${bright}${opt}${plain}"
			else
				# shellcheck disable=SC2206
				IFS=$'\n' lines=($opt)

				for line in "${lines[@]}"; do
					line="${line#"${line%%[![:space:]]*}"}"  # Remove leading whitespace
					echo "${indent}${indent}${line}"
				done
			fi
		done
	fi
}

# Shows an error message on stderr, and exits the script
function error() {
	echo -e "${light_red}Error: ${dark_red}$*${plain}" >&2
	if man -S1 -w "$script_name" &> /dev/null; then
		echo "Run '$script_name --help' for usage info, or 'man $script_name' for full details" >&2
	else
		echo "Run '$script_name --help' for usage info" >&2
	fi

	exit 1
}

# Errors unless all passed files/directories exist
# Symlinks to existing files/directories are allowed (the IDE will resolve them)
function error_unless_all_exist() {
	local allow=$1  # Pass one of: File, Directory, Path
	shift

	local arg=""
	for arg in "$@"; do
		if [[ ! -e $arg ]]; then
			if [[ -L $arg ]]; then
				error "Broken symlink: $arg"
			else
				error "$allow does not exist: $arg"
			fi
		elif [[ $allow == File ]]; then
			if [[ ! -f $arg ]]; then
				if [[ -L $arg ]]; then
					error "Symlink doesn't resolve to a regular file: $arg"
				else
					error "Not a regular file: $arg"
				fi
			fi
		elif [[ $allow == Directory ]]; then
			if [[ ! -d $arg ]]; then
				if [[ -L $arg ]]; then
					error "Symlink doesn't resolve to a directory: $arg"
				else
					error "Not a directory: $arg"
				fi
			fi
		elif [[ ! -f $arg && ! -d $arg ]]; then
			if [[ -L $arg ]]; then
				error "Symlink doesn't resolve to a regular file or directory: $arg"
			else
				error "Not a regular file or directory: $arg"
			fi
		fi
	done
}

# Errors if the IDE is already running
function error_if_running() {
	local command=$1  # e.g. format

	# shellcheck disable=SC2009  # pgrep can't match complete binary paths without also matching their arguments
	if ps -x -o comm= | grep -q "/${ide_name}[^/]*\.app"; then
		error "The $command command won't work right if $ide_name is already running"
	fi
}

# Launches the IDE via its MacOS binary (useful for some commands, not so much for others)
function exec_ide_binary() {
	local binary=""
	binary="$(cd "$ide_app/Contents/MacOS" &> /dev/null && echo *)"

	if [[ -z $binary || $binary == *" "* ]]; then
		binary=$ide_binary
	fi

	binary="$ide_app/Contents/MacOS/$binary"
	if [[ ! -f $binary || ! -x $binary ]]; then
		binary=""
	fi

	if [[ -z $binary ]]; then
		error "Unable to launch $ide_app: Can't find the right binary in its MacOS directory"
	fi

	exec "$ide_app/Contents/MacOS/$binary" "$@"
}

# Basic command-line parsing, just handling --wait and the weird no-dash global options
specials=" --nosplash --dontReopenProjects --disableNonBundledPlugins "
commands=" diff merge format inspect installPlugins "

wait=false
open_opts=()
special_opts=()
args=()
command=""

for arg; do
	if [[ $arg == --help || $arg == -help || $arg == -h ]]; then
		echo -e "${bright}${ide_name} launcher${plain}"
		echo -e "${dim}$ide_app${plain}"
		usage
		exit 0

	elif [[ $arg == --wait || $arg == -wait || $arg == -w ]]; then
		wait=true
		open_opts+=(-W)  # We need to pass an option to the "open" command too
		special_opts+=(--wait)

	elif [[ $specials == *" $arg "* || $specials == *" -$arg "* || $specials == *" --$arg "* ]]; then
		special_opts+=("${arg//-}")

	else
		args+=("$arg")
	fi
done

# No arguments, just launch (with any special options)
if [[ ${#args[@]} == 0 ]]; then
	if [[ $wait == true ]]; then
		error "You must pass a file or project directory with the --wait option"
	fi

# First argument is a command
elif [[ $commands == *" ${args[0]} "* ]]; then
	command=${args[0]}

	if [[ $wait == true ]]; then
		error "You cannot use the --wait option with the $command command"
	fi

	if [[ $command == diff ]]; then
		# Open the diff viewer to compare two or three files from the command line
		# https://www.jetbrains.com/help/idea/command-line-differences-viewer.html#d1a81073

		# We need 2 or 3 files; no options are allowed here
		if [[ ${#args[@]} != 3 && ${#args[@]} != 4 ]]; then
			error "You must pass the paths/names of 2 or 3 files to compare"
		else
			error_unless_all_exist File "${args[@]:1}"
		fi

	elif [[ $command == merge ]]; then
		# Open the Merge dialog to perform a three-way or a two-way merge from the command line
		# https://www.jetbrains.com/help/idea/command-line-merge-tool.html#macos

		# We need 3 or 4 files; no options are allowed here
		if [[ ${#args[@]} != 4 && ${#args[@]} != 5 ]]; then
			error "You must pass the paths/names of 2 versions of a file to merge," $'\n'\
				"optionally the base revision (a common origin of both modified versions)," $'\n'\
				"and then the output file to save merged results into."
		else
			error_unless_all_exist File "${args[@]:1}"
		fi

	elif [[ $command == format ]]; then
		# https://www.jetbrains.com/help/idea/command-line-formatter.html#9f540e09

		error_if_running format

		unset need_mask need_settings_path need_charset
		paths=()

		for arg in "${args[@]:1}"; do
			if [[ -n $need_mask ]]; then
				if [[ ! $arg =~ [*?] ]]; then
					error "Invalid mask used with the $need_mask option (no wildcards): $arg"
				else
					unset need_mask
				fi
			elif [[ -n $need_settings_path ]]; then
				error_unless_all_exist File "$arg"
				if [[ $arg != *.xml ]]; then
					error "The code style settings must be an XML file: $arg"
				else
					unset need_settings_path
				fi
			elif [[ -n $need_charset ]]; then
				unset need_charset  # We don't try to check this value

			elif [[ $arg == -m || $arg == -mask ]]; then
				need_mask=$arg
			elif [[ $arg == -s || $arg == -settings ]]; then
				need_settings_path=$arg
			elif [[ $arg == -charset ]]; then
				need_charset=$arg

			elif [[ $arg == -r || $arg == -R || $arg == -allowDefaults || $arg == -d || $arg == -dry ]]; then
				# Miscellaneous valid option, carry on
				continue
			elif [[ $arg == -* && ! -f $arg && ! -d $arg ]]; then
				error "Invalid option: $arg"
			else
				paths+=("$arg")
			fi
		done

		if [[ -n $need_mask ]]; then
			error "You must pass a file mask after the $need_mask option"
		elif [[ -n $need_settings_path ]]; then
			error "You must pass a path to a code style settings file after the $need_settings_path option"
		elif [[ -n $need_charset ]]; then
			error "You must pass a character set after the $need_charset option, e.g. ISO-8859-15"
		fi

		if [[ ${#paths[@]} == 0 ]]; then
			error "Nothing to format! (No files or directories passed)"
		else
			error_unless_all_exist Path "${paths[@]}"
		fi

		# Delegate to format.sh, if we can
		format_sh="$ide_app/Contents/bin/format.sh"

		if [[ -f $format_sh && -x $format_sh ]]; then
			exec "$format_sh" "${special_opts[@]}" "${args[@]:1}"
		else
			exec_ide_binary format "${special_opts[@]}" "${args[@]:1}"
		fi

	elif [[ $command == inspect ]]; then
		# https://www.jetbrains.com/help/idea/command-line-code-inspector.html#7d86c34f

		# Mostly you can put options anywhere on a command line, and the IDE will handle them as expected,
		# but for the inspect command, they must come at the end, or the command will fail,
		# so we accept them anywhere here, and manually move them to the end to keep the IDE happy

		error_if_running inspect

		unset need_subdir need_format profile
		dirs=()
		opts=()

		for arg in "${args[@]:1}"; do
			if [[ -n $need_subdir ]]; then
				error_unless_all_exist Directory "$arg"
				unset need_subdir
				opts+=("$arg")

			elif [[ -n $need_format ]]; then
				formats=" xml json plain "
				if [[ $formats != *" $arg "* ]]; then
					error "Invalid format used with $need_format: $arg"
				else
					unset need_format
					opts+=("$arg")
				fi
			
			elif [[ $arg == -d ]]; then
				need_subdir=$arg
				opts+=("$arg")
			elif [[ $arg == -format ]]; then
				need_format=$arg
				opts+=("$arg")
			
			elif [[ $arg == -changes || $arg =~ ^-v[012]$ ]]; then
				# Miscellaneous valid option, carry on
				opts+=("$arg")
			elif [[ $arg == -* && ! -f $arg && ! -d $arg ]]; then
				error "Invalid option: $arg"
			elif [[ ${#dirs[@]} == 1 && -z $profile ]]; then
				profile=$arg
			else
				dirs+=("$arg")
			fi
		done

		if [[ -n $need_subdir ]]; then
			error "You must pass a subdirectory path after the $need_subdir option"
		elif [[ -n $need_format ]]; then
			error "You must pass a format after the $need_format option (xml, json, or plain)"
		fi

		if [[ ${#dirs[@]} != 2 || -z $profile ]]; then
			error "Wrong number of arguments passed, check usage"
		elif [[ $profile != *.xml ]]; then
			error "The inspection profile must be an XML file: $profile"
		fi

		error_unless_all_exist Directory "${dirs[@]}"
		error_unless_all_exist File "$profile"

		# Rewrite $args with all options (and their values) at the end
		args=(inspect "${dirs[0]}" "$profile" "${dirs[1]}" "${opts[@]}")

		# Delegate to inspect.sh, if we can
		inspect_sh="$ide_app/Contents/bin/inspect.sh"

		if [[ -f $inspect_sh && -x $inspect_sh ]]; then
			exec "$inspect_sh" "${special_opts[@]}" "${args[@]:1}"
		else
			DEFAULT_PROJECT_PATH="$(pwd)"  # Copied from inspect.sh, unsure whether/why it's needed
			export DEFAULT_PROJECT_PATH
			exec_ide_binary inspect "${special_opts[@]}" "${args[@]:1}"
		fi

	elif [[ $command == installPlugins ]]; then
		# https://www.jetbrains.com/help/idea/install-plugins-from-the-command-line.html#macos

		error_if_running installPlugins
		exec_ide_binary installPlugins "${special_opts[@]}" "${args[@]:1}"
	fi

# Arguments are paths; the IDE will create files if they don't already exist,
# and add a .idea subdirectory when opening a directory that doesn't already have one.
# As of v2023.2.5: --line and its value are accepted, anywhere on the command line,
# including multiple times, but ignored AFAICT.
# https://www.jetbrains.com/help/idea/opening-files-from-command-line.html#macos
else
	need_line_num=false
	path_count=0

	for arg in "${args[@]}"; do
		if [[ $need_line_num == true ]]; then
			if [[ $arg =~ ^[0-9]+$ ]]; then
				need_line_num=false
			else
				break
			fi
		elif [[ $arg == --line ]]; then
			need_line_num=true
		else
			(( path_count+=1 ))
		fi
	done

	if [[ $need_line_num == true ]]; then
		error "You must pass a line number after the --line option"
	elif [[ $wait == true && $path_count != 1 ]]; then
		error "You must pass exactly one file or project directory with the --wait option"
	fi
fi

# Launch the IDE via 'open'
open -na "$ide_app" "${open_opts[@]}" --args "${special_opts[@]}" "${args[@]}"
